genie-web/src/main/java/com/netflix/genie/web/tasks/leader/DatabaseCleanupTask.java

```diff
@@ -411,19 +414,36 @@ private void deleteTags(final Instant creationThreshold) {
                     Integer.class,
                     this.cleanupProperties.getBatchSize()
                 );
-
+                final long rollingWindowHours = this.environment.getProperty(
+                    DatabaseCleanupProperties.TagDatabaseCleanupProperties.ROLLING_WINDOW_HOURS_PROPERTY,
+                    Integer.class,
+                    this.cleanupProperties.getTagCleanup().getRollingWindowHours()
+                );
+                final long batchDaysWithin = this.environment.getProperty(
+                    DatabaseCleanupProperties.TagDatabaseCleanupProperties.BATCH_DAYS_WITHIN_PROPERTY,
+                    Integer.class,
+                    this.cleanupProperties.getTagCleanup().getBatchDaysWithin()
+                );
                 log.info(
-                    "Attempting to delete unused tags from before {} in batches of {}",
+                    "Attempting to delete unused tags older than {} in batches of {}. Using a rolling window"
+                        + " of {} hours within last {} days",
                     creationThreshold,
-                    batchSize
+                    batchSize,
+                    rollingWindowHours,
+                    batchDaysWithin
                 );
 
-                long deleted;
                 long totalDeleted = 0L;
-                do {
-                    deleted = this.persistenceService.deleteUnusedTags(creationThreshold, batchSize);
-                    totalDeleted += deleted;
-                } while (deleted > 0);
+                Instant upperBound = creationThreshold;
+                Instant lowerBound = creationThreshold.minus(rollingWindowHours, ChronoUnit.HOURS);
+                final Instant batchLowerBound = creationThreshold.minus(batchDaysWithin, ChronoUnit.DAYS);
+                while (upperBound.isAfter(batchLowerBound)) {
+                    totalDeleted += deleteUnusedTagsBetween(lowerBound, upperBound, batchSize);
+                    upperBound = lowerBound;
+                    lowerBound = lowerBound.minus(rollingWindowHours, ChronoUnit.HOURS);
+                }
+                // do a final deletion of everything <= batchLowerBound
+                totalDeleted += deleteUnusedTagsBetween(Instant.EPOCH, upperBound, batchSize);
```

Comment:
from `epoch` to `upperBound` can be a large range, I think it's almost [epoch, today - 30 days), would this be a concern?

Reply:
It shouldn't be a concern, as the `batchSize` still limits the number of rows it processes and deletes in each iteration.